hail/src/main/scala/is/hail/shuffler/ShuffleClient.scala

```diff
@@ -1,99 +1,165 @@
+      Wire.writeByteArray(out, uuid)
+    }
+  }
 
   def start(): Unit = {
     log.info(s"CLNT start")
-    out.write(Wire.START)
-    Wire.writeTStruct(out, t)
-    Wire.writeTypedCodecSpec(out, codecSpec)
-    Wire.writeStringArray(out, key)
+    startOperation(Wire.START)
+    Wire.writeTStruct(out, rowType)
+    Wire.writeEBaseStruct(out, rowEType)
+    Wire.writeSortFieldArray(out, keyFields)
+    Wire.writeEBaseStruct(out, keyEType)
     out.flush()
     uuid = Wire.readByteArray(in)
+    assert(uuid.length == Wire.ID_SIZE, s"${uuid.length} ${Wire.ID_SIZE}")
     log.info(s"CLNT start done")
   }
 
-  def put(values: Iterator[Long]): Unit = {
+  def put(values: Array[Long]): Unit = {
     log.info(s"CLNT put")
-    val encoder = makeEnc(out)
-    encoder.writeByte(Wire.PUT)
-    encoder.flush()
-    Wire.writeByteArray(out, uuid)
+    startOperation(Wire.PUT)
     out.flush()
-    while (values.hasNext) {
-      encoder.writeByte(1)
-      val off = values.next
-      encoder.writeRegionValue(off)
-    }
-    encoder.writeByte(0)
-    encoder.flush()
-    assert(in.read() == 0)
+    val encoder = codecs.makeRowEncoder(out)
+    writeRegionValueArray(encoder, values)
+    // fixme: server needs to send uuid for the successful partition
+    out.flush()
+    assert(in.readByte() == 0.toByte)
     log.info(s"CLNT put done")
   }
 
-  def get(region: Region, start: Long, end: Long): Array[Long] = {
-    log.info(s"CLNT get")
-    val keyEncoder = makeKeyEnc(out)
-    val decoder = makeDec(in)
-    keyEncoder.writeByte(Wire.GET)
-    keyEncoder.flush()
-    Wire.writeByteArray(out, uuid)
+  def get(
+    region: Region,
+    start: Long,
+    startInclusive: Boolean,
+    end: Long,
+    endInclusive: Boolean
+  ): Array[Long] = {
+    log.info(s"CLNT get ${Region.pretty(codecs.keyDecodedPType, start)} ${startInclusive} " +
+      s"${Region.pretty(codecs.keyDecodedPType, end)} ${endInclusive}")
+    val keyEncoder = codecs.makeKeyEncoder(out)
+    val decoder = codecs.makeRowDecoder(in)
+    startOperation(Wire.GET)
     out.flush()
     keyEncoder.writeRegionValue(start)
+    keyEncoder.writeByte(if (startInclusive) 1.toByte else 0.toByte)
     keyEncoder.writeRegionValue(end)
+    keyEncoder.writeByte(if (endInclusive) 1.toByte else 0.toByte)
     keyEncoder.flush()
-
     log.info(s"CLNT get receiving values")
-    val ab = new ArrayBuilder[Long]()
-    var hasNext = decoder.readByte()
-    assert(hasNext >= 0)
-    while (hasNext == 1) {
-      ab += decoder.readRegionValue(region)
-      hasNext = decoder.readByte()
-      assert(hasNext >= 0)
-    }
-    assert(hasNext == 0)
+    val values = readRegionValueArray(region, decoder)
     log.info(s"CLNT get done")
-    ab.result()
+    values
+  }
+
+  def partitionBounds(region: Region, nPartitions: Int): Array[Long] = {
+    log.info(s"CLNT partitionBounds")
+    val keyDecoder = codecs.makeKeyDecoder(in)
+    startOperation(Wire.PARTITION_BOUNDS)
+    out.writeLong(nPartitions)
```

Comment:
eh, I was trying to make the protocol using longs everywhere because 32-bits seems so 1990. This probably is not necessary for n_partitions and anyway my key sampling is stored in a java array.